### PR TITLE
Remaking Appendix C to "Tips and Tricks"

### PR DESCRIPTION
**Pull request by rudeboybert**
Adding answers to the most commonly asked data wrangling questions we've encountered during student projects.

### CODE REVIEW

**Comment by rudeboybert** [comment]
Added the following page to Appendix C: https://rudeboybert.github.io/SDS192/tips_and_tricks.html

File: 93-appendixC.Rmd [hunk 1]
```diff
@@ -1,69 +1,351 @@
...
+
+While there are statistical methods to deal with missing data they are beyond the reach of this class. The easiest thing to do is to remove all missing cases, but **you should always at the very least report to the reader if you do so, as by removing the missing values you may be biasing your results.**
+
+You can do this with a `na.rm = TRUE` argument like so:
 
 ```{r}
-sorted_flights <- sort(flights_table, decreasing = TRUE)
-names(sorted_flights)
+movies %>% 
+  summarize(mean_profit = median(millions, na.rm = T))
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
  summarize(mean_profit = median(millions, na.rm = TRUE))
```

File: 93-appendixC.Rmd [hunk 2]
```diff
@@ -1,69 +1,351 @@
...
 ```
 
-```{r include=FALSE}
-#carrier_counts <- carrier_counts %>%
-#  arrange(desc(n))
+If you decide you want to remove the row with the missing data, you can use the filter function like so:
+
+```{r}
+movies <- movies %>%
```
Comment (ismayc) [inline 93-appendixC.Rmd]: Probably shouldn’t do a reassignment here but create a different variable instead?

File: 93-appendixC.Rmd [hunk 3]
```diff
@@ -1,69 +1,351 @@
...
+ggplot(revenue_by_type, aes(x = type, y = total_revenue)) +
+  geom_col() + 
+  labs(x = "Movie genre", y = "Total boxoffice revenue (in millions of $)")
 ```
 
-```{r include=FALSE}
-#ggplot(data = carrier_counts, mapping = aes(x = carrier, y = n)) +
-#  geom_bar(stat = "identity") + 
-#  scale_x_discrete(limits = carrier)
+Or if you want to reorder `type` in ascencding order of `total_revenue`, we use `reorder()`
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Or if you want to reorder `type` in ascending order of `total_revenue`, we use `reorder()`
```

File: 93-appendixC.Rmd [hunk 4]
```diff
@@ -1,69 +1,351 @@
...
+
+ggplot(revenue_by_type, aes(x = type, y = total_revenue)) +
+  geom_col() + 
+  labs(
+    x = "Movie genre", y = "Total boxoffice revenue (in millions of $)"
+  )
 ```
 
-The last addition here specifies the values of the horizontal `x` axis on a discrete scale to correspond to those given by the entries of `sorted_flights`.  
+Or if you want to reorder `type` in descencding order of `total_revenue`, just put
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Or if you want to reorder `type` in descending order of `total_revenue`, just put
```

File: 93-appendixC.Rmd [hunk 5]
```diff
@@ -1,69 +1,351 @@
...
+  geom_boxplot() + 
+  labs(x = "rating", y = "Profit in millions of $", title = "Profits for different movie ratings") +
+  scale_y_continuous(labels = dollar)
+```
+
+
+
+### Changing values inside cells
+
+`dplyr::rename()` renames column/variable names. To "rename" values inside cells of a particular column, you need to `mutate()` the column using one of the three functins below. There might be other ones too, but these are the three I've seen the most. In these examples, we'll change values in the variable `type`.
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
`dplyr::rename()` renames column/variable names. To "rename" values inside cells of a particular column, you need to `mutate()` the column using one of the three functions below. There might be other ones too, but these are the three I've seen the most. In these examples, we'll change values in the variable `type`.
```

File: 93-appendixC.Rmd [hunk 6]
```diff
@@ -1,69 +1,351 @@
...
+
+1. `ifelse()`
+1. `recode()`
+1. `case_when()`
+
+#### `ifelse()` {-}
```
Comment (ismayc) [inline 93-appendixC.Rmd]: Probably best to switch this to use `dplyr::if_else()` instead here

File: 93-appendixC.Rmd [hunk 7]
```diff
@@ -1,69 +1,351 @@
...
+             # Need this for everything else that aren't the two cases above:
+             TRUE ~ "Rest"
+           ))
+```
+
+
+
+### Converting a numerical variable to a categorical one
+
+Sometimes we want to turn a numerical, continuous variable into a categorical variable. For instance, what if we wanted to have a variable that tells us if a movie made hundred million dollars or more. In other words a binary variable in other words a categorical variable with 2 levels. We can again use the `mutate` function:
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Sometimes we want to turn a numerical, continuous variable into a categorical variable. For instance, what if we wanted to have a variable that tells us if a movie made one hundred million dollars or more. In other words a binary variable in other words a categorical variable with 2 levels. We can again use the `mutate` function:
```

Comment (ismayc) [reply 93-appendixC.Rmd]: Can this be rewritten to not use “in other words”?

File: 93-appendixC.Rmd [hunk 8]
```diff
@@ -1,69 +1,351 @@
...
+```
+
+What if you want to convert a numerical variable into a categorical variable with more than 2 levels? One way is to use the `cut()` command. For instance, below, we `cut()` the `score` variable, to recode it into 4 categories:
+
+1. 0 - 40 = bad
+2. 40.1 - 60 = so-so
+3. 60.1 - 80 = good
+4. 80.1+ = great
+
+We set the breaking points for cutting the numerical variable with the `c(0, 40, 60, 80, 100)` part, and set the labels for each of these bins with the `labels=c("bad","so-so","good","great")` part. All this action happens inside the `mutate` command, so the new categorical variable `scorecats` is added to the data frame.
```
Comment (ismayc) [inline 93-appendixC.Rmd]: `score_categ` instead here

File: 93-appendixC.Rmd [hunk 9]
```diff
@@ -1,69 +1,351 @@
...
+  mutate(score_categ = cut(score, 
+                           breaks = c(0, 40, 60, 80, 100),
+                           labels = c("bad", "so-so", "good", "great")
+  )) %>% 
+  slice(1:10)
+```
+
+Other options with the `cut` function:
+
+*  By default, if the value is exacly the upper bound of an interval, it's
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
*  By default, if the value is exactly the upper bound of an interval, it's
```

File: 93-appendixC.Rmd [hunk 10]
```diff
@@ -1,69 +1,351 @@
...
+included in the lessor category (e.g. 60.0 is 'so-so' not 'good'), to 
+flip this, include the argument `right = FALSE`.
+*  You could also have Rstudio equally divide the variable into a balanced
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
*  You could also have RStudio equally divide the variable into a balanced
```

File: 93-appendixC.Rmd (hunk 2 above)
Comment (rudeboybert) [reply 93-appendixC.Rmd]: yep!

File: 93-appendixC.Rmd (hunk 6 above)
Comment (rudeboybert) [reply 93-appendixC.Rmd]: sounds good

File: 93-appendixC.Rmd (hunk 8 above)
Comment (rudeboybert) [reply 93-appendixC.Rmd]: sounds good

**Comment by rudeboybert** [comment]
My bad on not having run the spell-check. I've already forgotten that part of the routine!

File: 93-appendixC.Rmd [hunk 11]
```diff
@@ -1,69 +1,351 @@
...
 library(nycflights13)
 ```
 
-## Sorted barplots
+## Data wrangling
+
+> Shout out to [Dr. Jenny Smetzer](https://www.scsparkscience.org/fellow/jennifer-smetzer/) for her work setting this up!
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, message=FALSE, warning = FALSE, eval = TRUE,
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, eval = TRUE, 
```

File: 93-appendixC.Rmd [hunk 12]
```diff
@@ -1,69 +1,351 @@
...
+                      fig.width = 16/2, fig.height = 9/2)
+
+# Why 76? https://www.youtube.com/watch?v=xjJ7FheCkCU
+set.seed(76)
+```
+
+![](images/data_ninja1.png){width=30%}      
 
-Building upon the example in Section \@ref(geombar):
+Let's load a movies dataset, pare down the rows and columns a bit, and then show the first 10 rows using `slice()`
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Let's load a movies dataset, pare down the rows and columns a bit, and then show the first 10 rows using `slice()`.
```

File: 93-appendixC.Rmd [hunk 13]
```diff
@@ -1,69 +1,351 @@
...
+  geom_boxplot() + 
+  labs(x = "rating", y = "Profit in millions of $", title = "Profits for different movie ratings") +
+  scale_y_continuous(labels = dollar)
+```
+
+
+
+### Changing values inside cells
+
+`dplyr::rename()` renames column/variable names. To "rename" values inside cells of a particular column, you need to `mutate()` the column using one of the three functions below. There might be other ones too, but these are the three I've seen the most. In these examples, we'll change values in the variable `type`.
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
`dplyr::rename()` renames column/variable names. To "rename" values inside cells of a particular column, you need to `mutate()` the column using one of the three functions below. There might be other ones too, but these are the three we've seen the most. In these examples, we'll change values in the variable `type`.
```

File: 93-appendixC.Rmd [hunk 14]
```diff
@@ -1,69 +1,351 @@
...
+             # Need this for everything else that aren't the two cases above:
+             TRUE ~ "Rest"
+           ))
+```
+
+
+
+### Converting a numerical variable to a categorical one
+
+Sometimes we want to turn a numerical, continuous variable into a categorical variable. For instance, what if we wanted to have a variable that tells us if a movie made one hundred million dollars or more. In other words a binary variable in other words a categorical variable with 2 levels. We can again use the `mutate` function:
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Sometimes we want to turn a numerical, continuous variable into a categorical variable. For instance, what if we wanted to have a variable that tells us if a movie made one hundred million dollars or more. That is to say, we can create a binary variable, which is the same thing as a categorical variable with 2 levels. We can again use the `mutate()` function:
```

File: 93-appendixC.Rmd [hunk 15]
```diff
@@ -1,69 +1,351 @@
...
+```
+
+What if you want to convert a numerical variable into a categorical variable with more than 2 levels? One way is to use the `cut()` command. For instance, below, we `cut()` the `score` variable, to recode it into 4 categories:
+
+1. 0 - 40 = bad
+2. 40.1 - 60 = so-so
+3. 60.1 - 80 = good
+4. 80.1+ = great
+
+We set the breaking points for cutting the numerical variable with the `c(0, 40, 60, 80, 100)` part, and set the labels for each of these bins with the `labels=c("bad","so-so","good","great")` part. All this action happens inside the `mutate` command, so the new categorical variable `score_categ` is added to the data frame.
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
We set the breaking points for cutting the numerical variable with the `c(0, 40, 60, 80, 100)` part, and set the labels for each of these bins with the `labels = c("bad", "so-so", "good", "great")` part. All this action happens inside the `mutate()` command, so the new categorical variable `score_categ` is added to the data frame. 
```

File: 93-appendixC.Rmd [hunk 16]
```diff
@@ -1,69 +1,351 @@
...
+  slice(1:10)
+```
+
+Other options with the `cut` function:
+
+*  By default, if the value is exactly the upper bound of an interval, it's
+included in the lessor category (e.g. 60.0 is 'so-so' not 'good'), to 
+flip this, include the argument `right = FALSE`.
+*  You could also have RStudio equally divide the variable into a balanced
+number of groups.  For example, specifying `breaks=3` would create 4 groups with
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
number of groups.  For example, specifying `breaks = 3` would create 4 groups with
```

File: 93-appendixC.Rmd [hunk 17]
```diff
@@ -1,69 +1,351 @@
...
+```
+
+So for example, the 4 proportions corresponding to R rated movies are 0.596 + 0.142 + 0.213 + 0.0491 = 1.
+
+
+
+
+### Dealing with %, commas, and \$ {#numerical}
+
+Say you have numerical data that are recorded as percentages, have commas, or are in dollar form and hence are character strings. How do you convert these to numerical values? Using `readr::parse_number` inside a `mutate()`! Shout out to [Stack Overflow](https://stackoverflow.com/a/52965808/3674500){target="_blank"}
```
Comment (ismayc) [inline 93-appendixC.Rmd]:
```suggestion
Say you have numerical data that are recorded as percentages, have commas, or are in dollar form and hence are character strings. How do you convert these to numerical values? Using `readr::parse_number()` inside a `mutate()`! Shout out to [Stack Overflow](https://stackoverflow.com/a/52965808/3674500){target="_blank"}
```

**Comment by rudeboybert** [comment]
Absolutely. I'm a little harried b/c the next phase of my students' project is coming real quick. Thanks for the fine-level edits

**Comment by ismayc** [comment]
Just waiting for travis to finish and then I'll merge this in.

**Comment by ismayc** [comment]
Feel free to delete the branch if you are done with it.

**Pull request merged**